src/vtl.ts

```diff
@@ -418,4 +437,78 @@ export class VTL {
     const __exhaustive: never = node;
     return __exhaustive;
   }
+
+  private mapForEachArgs(call: CallExpr) {
+    const fn = call.args.callbackfn as FunctionExpr;
+    return fn.parameters.map((p) => p.name);
+  }
+
+  private renderMapOrForEach(
+    call: CallExpr,
+    list: string,
+    // Should start with $
+    returnVariable?: string,
+    index?: string,
+    array?: string
+  ) {
+    if (index) {
+      this.add(`#set($${index} = $foreach.index)`);
+    }
+    if (array) {
+      this.add(`#set($${array} = ${list})`);
+    }
+
+    const fn = call.args.callbackfn as FunctionExpr;
+
+    const tmp = returnVariable ? returnVariable : this.newLocalVarName();
+
+    for (const stmt of fn.body.statements) {
+      this.eval(stmt, tmp);
+    }
+
+    return tmp;
+  }
+
+  /**
+   * Recursively dives into map operations until a non-map or a map with arr paremeter is found.
+   * evaluates the expression after the last map.
+   *
+   * @return [firstVariable, list variable, render function]
+   */
+  private divePropAccessMapOperations(
+    expr: Expr,
+    // Should start with $
+    returnVariable: string,
+    alwaysEvaluate?: boolean
+  ): [string, string, () => void] {
+    if (
+      !alwaysEvaluate &&
+      expr.kind === "CallExpr" &&
+      expr.expr.kind === "PropAccessExpr" &&
+      expr.expr.name === "map"
+    ) {
+      const [value, index, array] = this.mapForEachArgs(expr);
+
+      const next = expr.expr.expr;
+
+      // If we find array, stop recursing as array wouldn't be correct in the current map.
+      const [lastValue, list, lastRender] = this.divePropAccessMapOperations(
+        next,
+        `$${value}`,
+        !!array
```

Comment:
What does this achieve?

Reply:
Will document. A map, reduce, or forEach that uses `array` cannot be simplified because the array wouldn't be accurate to the point in time as `array` is the previous array, which when flattened, has yet to be computed.